FeatureExtractor.predict() can now return labels in addition to features

cofri opened the pull request:
A new bool argument `return_labels` is introduced in `FeatureExtractor.predict()`. If True, the function returns a tuple `(features, labels)`. If False, the function behaves as previously: only the features are returned as a tensor.

Special cases:
- if no label is given in the dataset, `None` is returned
- if labels are one-hot encoded, they are reduced to single values.

paulnovello commented on oodeel/extractor/keras_feature_extractor.py:
```diff
@@ -164,22 +164,33 @@ def predict_tensor(self, tensor: TensorType) -> Union[tf.Tensor, List[tf.Tensor]
-            return self.predict_tensor(tensor)
+            features = self.predict_tensor(tensor)
+            if return_labels:
+                return features, None
```
Should be features, None if `isinstance(dataset, tf.Tensor)` and features, label otherwise. The problem is: how to identify label in dataset item? dataset[-1]? dataset['label']? (Maybe [1] to be consistent with the rest of your code)

paulnovello commented on oodeel/extractor/keras_feature_extractor.py:
```diff
@@ -190,9 +201,28 @@ def predict(
 
+            # Get labels
+            if return_labels and contains_labels:
+                lbl_batch = elem[1]  # labels must be at index 1 in the batch tuple
+                # If labels are one-hot encoded, take the argmax
+                if tf.rank(lbl_batch) > 1 and lbl_batch.shape[1] > 1:
+                    lbl_batch = tf.reshape(lbl_batch, shape=[lbl_batch.shape[0], -1])
+                    lbl_batch = tf.argmax(lbl_batch, axis=1)
+                # If labels are in two dimensions, squeeze them
+                if len(lbl_batch.shape) > 1:
+                    lbl_batch = tf.reshape(lbl_batch, [lbl_batch.shape[0]])
```
Maybe make a function out of it and use it twice, here and in the case of  `isinstance(dataset, get_args(ItemType)) == True`

y-prudent reviewed (changes requested):
Great work!
Mahalanobis code could be updated with this feature

paulnovello commented on oodeel/extractor/torch_feature_extractor.py:
```diff
@@ -224,9 +231,15 @@ def predict(
             tensor = TorchDataHandler.get_input_from_dataset_item(dataset)
-            return self.predict_tensor(tensor, detach=detach)
+            features = self.predict_tensor(tensor, detach=detach)
+            if return_labels:
```
In that case, we should be able to also return the label found in the given item, if there is any. Should be features, None if `isinstance(dataset, tf.Tensor)` and features, label otherwise. The problem is: how to identify label in dataset item? dataset[-1]? dataset['label']? (Maybe [1] to be consistent with the rest of your code)

paulnovello commented on oodeel/extractor/torch_feature_extractor.py:
```diff
@@ -237,9 +250,28 @@ def predict(
 
+            # Get labels
+            if return_labels and contains_labels:
+                lbl_batch = elem[1]  # labels must be at index 1 in the batch tuple
+                # If labels are one-hot encoded, take the argmax
+                if len(lbl_batch.shape) > 1 and lbl_batch.shape[1] > 1:
+                    lbl_batch = lbl_batch.view(lbl_batch.size(0), -1)
+                    lbl_batch = torch.argmax(lbl_batch, dim=1)
+                # If labels are in two dimensions, squeeze them
+                if len(lbl_batch.shape) > 1:
+                    lbl_batch = lbl_batch.view([lbl_batch.shape[0]])
```
Maybe make a function out of it and use it twice, here and in the case of  `isinstance(dataset, get_args(ItemType)) == True`

paulnovello reviewed (changes requested):
See comments therein. GG for the tests!!

y-prudent commented on oodeel/methods/mahalanobis.py:
```diff
@@ -58,23 +58,7 @@ def _fit_to_dataset(self, fit_dataset: DatasetType) -> None:
-            features.append(_features)
-        labels = self.op.cat(labels)
-        features = self.op.cat(features)
+        features, labels = self.feature_extractor.predict(fit_dataset)
```
You forgot the argument `return_labels=True`

cofri commented:
I've changed `.predict()` of the feature extractors to handle a list/tuple of tensors [data, labels]. Also added a unit test to assert this feature.
Note that the object `list` is now an `ItemType`. This is required because a PyTorch batch `next(iter(dataloader))` returns a list of `[data, labels]`, compared to TensorFlow which returns a tuple `(data, labels)`. 

paulnovello reviewed (approved):
LGTM
